### PR TITLE
Fixes and config files to reproduce SRD

### PR DESCRIPTION
**Pull request by fjaviersanchez**

### CODE REVIEW

**Comment by fjaviersanchez** [comment]
This PR addresses #10 and #11, building upon #12

**Comment by c-d-leonard** [comment]
Is this ready to merge? me and @nikosarcevic are playing with augur / firecrown for some forecasts and it would be great to have these updates to generate.py in master. I could review but I'm not that familar with augur - @slosar ?

**Comment by fjaviersanchez** [comment]
Thanks @c-d-leonard! I think this is ready to go, but I'd appreciate a look before merging.

File: augur/generate.py
Comment (slosar) [inline augur/generate.py:161]: Could you do here something like `scipy.ndimage.gaussian_filter` or something?

File: augur/generate.py
Comment (slosar) [inline augur/generate.py:159]: Ok, no, I think if Nz_type is Gaussian, then it should be Gaussian, not SRD.
We need a NZ_type == "SRD2018" or something that will be the SRD assumed. Others should do exactly what is says on the can (i.e. we want to use this also for things not in SRD)

File: augur/generate.py
Comment (slosar) [inline augur/generate.py:169]: See comment above. Tophat is tophat.

File: augur/generate.py
Comment (slosar) [inline augur/generate.py:329]: All this should be replaced with TJPCov (when it is ready!)

File: examples/srd_v1_3x2.yaml
Comment (slosar) [inline examples/srd_v1_3x2.yaml:29]: Ok, these biases should be 0.95/growth. Can you check that they are indeed this? And there should be a nice way for augur to calculate those itself. Perhaps similar to how I enabled np.logspace to be used for ell edges?

File: examples/srd_v1_3x2.yaml
Comment (slosar) [inline examples/srd_v1_3x2.yaml:73]: Again, Nz_type should be SRD here.

File: examples/srd_y10_3x2.yaml
Comment (slosar) [inline examples/srd_y10_3x2.yaml:1]: Is it true that `srd_y10_3x2.yaml` and `srd_y1_3x2.yaml` only differ in the number of source? Is there a way to avoid too much boilerplate? Ask becker if you don't know.

File: examples/srd_y10_3x2.yaml
Comment (slosar) [inline examples/srd_y10_3x2.yaml:72]: Again, some automated way to specify these 2.1, 3.2, etc would be nice. Something that user does not need to calculate them manually every time...

**Review by slosar — CHANGES_REQUESTED** [review]
I made some comments, but it would be good if a second person had a look.

**Comment by slosar** [comment]
> Is this ready to merge? me and @nikosarcevic are playing with augur / firecrown for some forecasts and it would be great to have these updates to generate.py in master. I could review but I'm not that familar with augur - @slosar ?

@c-d-leonard Danielle, if you could review as well that would be great. I had a quick look, but it is 10PM here and time to go to bed. But I'm super stoked to see this shit moving forward.

File: examples/srd_y10_3x2.yaml
Comment (fjaviersanchez) [reply examples/srd_y10_3x2.yaml:1]: They differ in the number of lens bins too. I'll ask Matt and see if there's anything we can do to avoid so much repetition.

File: augur/generate.py
Comment (fjaviersanchez) [reply augur/generate.py:161]: I checked and there are a couple of caveats. The first one is that it doesn't allow you to use a kernel with varying sigma (in this case I'm doing `sigma = sig_z*(1+z)`). The second problem is that if you just go and set `sigma = 0.05*(1+zcent)` for the Gaussian filter it won't work. The filter sigma has to be specified in units of the step in the `zall` array that I'm using there (so itd' be `0.05*(1+zcent)/(zall[1]-zall[0])`). It is true that the convolution is ~30 times faster but I think this is a one time operation (and also, it only takes 30 ms in the slow version).

File: augur/generate.py
Comment (c-d-leonard) [reply augur/generate.py:159]: Agree with Anze here.

File: augur/generate.py
Comment (c-d-leonard) [inline augur/generate.py:172]: I think this would also be "ConstantDensSRD2018" or something by the suggested naming convention above.

File: examples/srd_v1_3x2.yaml
Comment (c-d-leonard) [inline examples/srd_v1_3x2.yaml:18]: is there a reason to use eisenstein_hu here rather than camb or class? just curious

File: examples/srd_v1_3x2.yaml
Comment (c-d-leonard) [reply examples/srd_v1_3x2.yaml:29]: I can help with this, because I remember banging my head against a wall trying to figure it out a couple years ago - they aren't 0.95 / growth. There was a mistake in the calculation of the bias values that got used in the SRD, something to do with normalisation, which is why these are "magic numbers from Tim" i.e. the values with this mistake, which you have to use if you want to match the SRD. In general, when we are not trying to match the SRD, calculating 0.95 / growth makes sense to me.

**Review by c-d-leonard — CHANGES_REQUESTED** [review]
Looks good in general, I think the main thing is making separate SRD type labels for dNdz. Let me know if you want me to tag in and try to do that if you don't have bandwidth right now.

File: examples/srd_v1_3x2.yaml
Comment (fjaviersanchez) [reply examples/srd_v1_3x2.yaml:18]: It was to avoid problems with CAMB for certain combinations of w0-wa

**Comment by fjaviersanchez** [comment]
Thanks for taking a look @c-d-leonard and for offering to help! Yes, I also had some back and forth with the bias mismatch that you point out. I think that Tim's magic numbers correspond to 1.33/growth (which makes me think that there's a sigma8 factor missing somewhere). So your suggestion would be to just use 0.95/growth in the config, right? I think I can take care of these updates today, but if you want to take a stab at something in particular, please let me know! Thanks again!

File: augur/generate.py
Comment (slosar) [reply augur/generate.py]: Fine, then leave it as it is.

File: examples/srd_v1_3x2.yaml
Comment (slosar) [reply examples/srd_v1_3x2.yaml:29]: Ok, it would be good to see how big the difference this is. And that what you are describing doesn't get lost in time. Basically, when we update SRD and the numbers are different, we can say: they are this different due to different bias assumptions, and this different due to better modelling of whatever, etc

**Comment by slosar** [comment]
So the bias could be 0.95/g or 1.05/g not sure, but 1.33/g is def too much (making numbers look better than they should be). What does the SRD document actually say?

**Comment by c-d-leonard** [comment]
The SRD says b(z) = 0.95 / G(z) for Y10 and b(z) = 1.05 / G(z) for Y1 (section D1.1, page 47). I also just checked the slack history of my chat with Tim about this and he said that the SRD values used the "unnormalised growth" which I took to mean there was an error in their calculation where the growth just didn't get normalised at all - we could probably reverse-engineer what b(z) = Num / G(z) relation that is equivalent to if we want to.

File: examples/srd_v1_3x2.yaml
Comment (c-d-leonard) [reply examples/srd_v1_3x2.yaml:29]: Is there a good place to make note of this so it's not lost in time?

**Comment by c-d-leonard** [comment]
I'm happy with the way it is currently to merge in - is there anything else you want changed before merging @slosar? I feel like the issue of what are the bias values is kind of a separate question.

**Review by slosar — APPROVED** [review]

**Comment by slosar** [comment]
If it is good with Danielle it is good with me.
I'll let you guys merge.
On bias stuff: I've spoken to Rachel and it is known. So one of our first exercises should be to compare SRD with wrong biases vs SRD with the right biases.

**Comment by slosar** [comment]
I'll let Javi have to Joy of pressing merge, it always give me kicks.

**Pull request merged**
